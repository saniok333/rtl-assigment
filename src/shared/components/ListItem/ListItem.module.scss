@import '../../styles/mixins.scss';

.itemContainer {
  width: 800px;
  height: 500px;
  margin: 50px;
  box-shadow: 5px 5px 20px black;
  border-radius: 5px;
  overflow: hidden;
  position: relative;

  @include breakpoint(1400px, max) {
    width: 500px;
    height: 350px;
  }

  @include breakpoint(1200px, max) {
    width: 350px;
    height: 200px;
  }

  @include breakpoint(768px, max) {
    width: 200px;
    height: 150px;
  }
  .infoContainer {
    height: 30%;
    width: 100%;
    padding: 6px;
    box-sizing: border-box;
    position: absolute;
    bottom: 0;
    background: rgb(27, 27, 27, 0.5);
    color: white;
    transition: 0.5s;
    border-radius: 5px;
    @include breakpoint(1400px, max) {
      height: 55%;
    }
    .label {
      margin: 10px;
      font-size: 30px;
      color: #f80;
      @include breakpoint(1200px, max) {
        font-size: 18px;
      }
      @include breakpoint(768px, max) {
        margin: 5px;
        font-size: 12px;
      }
    }
    .title {
      margin: 10px;
      font-size: 40px;
      @include breakpoint(1200px, max) {
        font-size: 20px;
      }
      @include breakpoint(768px, max) {
        margin: 5px;
        font-size: 15px;
      }
    }
    .description {
      margin: 20px;
      font-size: 20px;
      opacity: 0;
      transition: all 0.2s;
      @include breakpoint(1200px, max) {
        font-size: 10px;
      }
      @include breakpoint(768px, max) {
        font-size: 8px;
        margin: 10px;
      }
    }
  }
  &:hover {
    cursor: pointer;
    .infoContainer {
      height: 100%;
      background: black;
      .description {
        opacity: 1;
      }
    }
  }
}
