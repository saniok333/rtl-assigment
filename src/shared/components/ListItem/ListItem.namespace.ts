import { BREAK_POINTS } from '../../constants/breakPoints';
import { SrcMapTypes } from '../../utils/types/srcMap.types';

export const ListItemSrcMap = {
  [BREAK_POINTS.EXTRA_LARGE]: {
    type: SrcMapTypes.liggend,
  },
  [BREAK_POINTS.LARGE]: {
    type: SrcMapTypes.artikelTop,
  },
  [BREAK_POINTS.MEDIUM]: {
    type: SrcMapTypes.inlineSmall,
  },
};
