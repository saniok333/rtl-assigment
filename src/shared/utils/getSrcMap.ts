import { ISrcMapItem } from './types/srcMap.types';

interface ISrcMap {
  [key: number]: { type: string };
}

interface IGetSrcMap {
  config: ISrcMap;
  crops: ArticlesModel.Crop[];
}

export const getSrcMap = ({ config, crops }: IGetSrcMap): ISrcMapItem[] => {
  const pictureConfig = Object.entries(config).map(([breakPoint, { type }]) => {
    const src = crops.find((imageCrop) => imageCrop.type === type)?.path ?? '';

    return {
      breakPoint,
      src,
    };
  });

  return pictureConfig.filter((item) => item.src);
};
