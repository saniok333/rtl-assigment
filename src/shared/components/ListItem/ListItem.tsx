import React, { useRef } from 'react';
import { useOnScreen } from '../../hooks/useOnScreen';
import { getSrcMap } from '../../utils/getSrcMap';
import { Picture } from '../Picture';
import styles from './ListItem.module.scss';
import { ListItemSrcMap } from './ListItem.namespace';

interface IImageData {
  afbeelding: string;
  crops: ArticlesModel.Crop[];
}

export interface PropsT {
  title: string;
  label: string;
  description: string;
  urlAlias: string;
  imageData: IImageData;
}

const ListItem: React.FC<PropsT> = ({
  title,
  label,
  description,
  urlAlias,
  imageData,
}): JSX.Element => {
  const { crops, afbeelding } = imageData;

  const containerRef = useRef<HTMLDivElement | null>(null);
  const isVisible = useOnScreen(containerRef);

  const srcMap = getSrcMap({
    config: ListItemSrcMap,
    crops,
  });

  return (
    <div ref={containerRef} className={styles.itemContainer}>
      <a href={urlAlias}>
        {isVisible && <Picture srcMap={srcMap} defaultSrc={afbeelding} />}
        <div className={styles.infoContainer}>
          <h3 className={styles.label}>{label}</h3>
          <h2 className={styles.title}>{title}</h2>
          <p className={styles.description}>{description}</p>
        </div>
      </a>
    </div>
  );
};

export default ListItem;
