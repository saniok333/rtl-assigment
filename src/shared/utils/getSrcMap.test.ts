import { ListItemSrcMap } from '../components/ListItem/ListItem.namespace';
import { getSrcMap } from './getSrcMap';

const cropsMock: ArticlesModel.Crop[] = [
  {
    type: 'inline_small',
    height: 150,
    width: 200,
    ratio: '1.33:1',
    path: 'https://www.rtlboulevard.nl/sites/default/files/styles/inline_small/public/content/images/2018/06/19/ANP-56919532.jpg?itok=uYTfDZPr',
  },
  {
    type: 'artikel_top',
    height: 395,
    width: 694,
    ratio: '1.76:1',
    path: 'https://www.rtlboulevard.nl/sites/default/files/styles/artikel_top/public/content/images/2018/06/19/ANP-56919532.jpg?itok=SikI_AJI',
  },
  {
    type: 'liggend',
    height: 576,
    width: 1024,
    ratio: '1.78:1',
    path: 'https://www.rtlboulevard.nl/sites/default/files/styles/liggend/public/content/images/2018/06/19/ANP-56919532.jpg?itok=TBK54eRq',
  },
  {
    type: 'liggend_breed_nieuwsopening',
    height: 430,
    width: 1024,
    ratio: '2.38:1',
    path: 'https://www.rtlboulevard.nl/sites/default/files/styles/liggend_breed_nieuwsopening/public/content/images/2018/06/19/ANP-56919532.jpg?itok=HUPFexR0',
  },
  {
    type: 'square_medium',
    height: 400,
    width: 400,
    ratio: '1:1',
    path: 'https://www.rtlboulevard.nl/sites/default/files/styles/square_medium/public/content/images/2018/06/19/ANP-56919532.jpg?itok=TE8J4FE3',
  },
];

describe('getSrcMap', () => {
  it('should return srcMap', () => {
    const extraPath1 = cropsMock[3].path;
    const extraPath2 = cropsMock[4].path;

    const srcMap = getSrcMap({
      config: ListItemSrcMap,
      crops: cropsMock,
    });

    expect(srcMap).toMatchSnapshot();
    expect(srcMap.find((item) => item.src === extraPath1)).toBeFalsy();
    expect(srcMap.find((item) => item.src === extraPath2)).toBeFalsy();
  });

  it('should skip 1200 breakpoint when missing path', () => {
    cropsMock[1].path = '';

    const srcMap = getSrcMap({
      config: ListItemSrcMap,
      crops: cropsMock,
    });

    expect(srcMap).toMatchSnapshot();
  });
});
