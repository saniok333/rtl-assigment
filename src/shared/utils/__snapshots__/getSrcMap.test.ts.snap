// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSrcMap should return srcMap 1`] = `
Array [
  Object {
    "breakPoint": "768",
    "src": "https://www.rtlboulevard.nl/sites/default/files/styles/inline_small/public/content/images/2018/06/19/ANP-56919532.jpg?itok=uYTfDZPr",
  },
  Object {
    "breakPoint": "1200",
    "src": "https://www.rtlboulevard.nl/sites/default/files/styles/artikel_top/public/content/images/2018/06/19/ANP-56919532.jpg?itok=SikI_AJI",
  },
  Object {
    "breakPoint": "1400",
    "src": "https://www.rtlboulevard.nl/sites/default/files/styles/liggend/public/content/images/2018/06/19/ANP-56919532.jpg?itok=TBK54eRq",
  },
]
`;

exports[`getSrcMap should skip 1200 breakpoint when missing path 1`] = `
Array [
  Object {
    "breakPoint": "768",
    "src": "https://www.rtlboulevard.nl/sites/default/files/styles/inline_small/public/content/images/2018/06/19/ANP-56919532.jpg?itok=uYTfDZPr",
  },
  Object {
    "breakPoint": "1400",
    "src": "https://www.rtlboulevard.nl/sites/default/files/styles/liggend/public/content/images/2018/06/19/ANP-56919532.jpg?itok=TBK54eRq",
  },
]
`;
